# ¡Feliz Navidad a toda LATAM! 🎄✨

Soy **Hesus Garcia** – tu revisor de confianza y ahora también el **Hesus que trae la magia navideña al código** 🎅🦸‍♂️. Sí, soy el único Hesus con H que viene cargado de observaciones, sugerencias y buenos deseos para ayudarte a mejorar tu proyecto en estas fechas tan especiales. 🎁

Quiero aprovechar esta oportunidad para desearte a ti y a toda nuestra increíble comunidad de estudiantes en Latinoamérica unas fiestas llenas de alegría, aprendizaje y nuevos logros. 🌟 ¡Que esta Navidad esté llena de código limpio, proyectos exitosos y muchos "outputs" perfectos! 🚀

Cada vez que encuentre un detalle importante en tu código, te lo señalaré con la intención de que puedas corregirlo y brilles en tu camino profesional. Recuerda que estos comentarios no son más que regalos disfrazados, pensados para ayudarte a crecer y mejorar. 🎁 Si alguna vez te encuentras con un problema que no logras resolver, no te preocupes; estaré aquí para guiarte.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**. Estos son como los adornos de un árbol de Navidad 🎄; cada uno está puesto en su lugar por una razón. 

---

### **Formato de Comentarios**

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:

---

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones. 🎉
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto. 🔧
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados. 🚨
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor**, como un resumen festivo de los aspectos positivos y áreas para mejorar en tu proyecto. Esto incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto. 🎯
- **Áreas de mejora:** Sugerencias donde puedes mejorar. 🔍
- **Temas adicionales para investigar:** Ideas opcionales para expandir tus habilidades. 🌱

---

Espero que este espíritu navideño te motive a seguir adelante y que este proyecto sea uno de muchos logros en tu camino. ¡Felicidades por tu dedicación y esfuerzo! 💪🎅✨

### ¡Empecemos con esta revisión navideña! 🎄🚀

# **Proyecto 10**

**Tienda online**

Junto con el departamento de marketing se ha recopilado una lista de hipótesis que pueden ayudar a aumentar los ingresos de la tienda.

Para este proyecto primero se priorizarán las hipótesis, para elegir la hipótesis más adecuada; se realizará un test A/B, para posteriormente analizar los resultados.

**Librerías**

In [5]:
# Importo todas las librerías que me serán de utilidad a lo largo del proyecto.

import pandas as pd
import numpy as np
import scipy.stats as stats
import math as mth
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px


**Preparación de los datos: Dataframes (df)**

In [6]:
# Leo los archivos csv, los convierto en dataframes y me familiarizo con los datos que contienen.

# El df hypotheses contiene las hipótesis que se ponderarán.
# Sus columnas son calificaciones de las escalas Reach, Impact, Confidence, y Effort. 
hypotheses = pd.read_csv('./datasets/hypotheses_us.csv', sep=';')

# El df orders son las ordenes hechas por los usuarios de ambos grupos (A/B)
orders = pd.read_csv('./datasets/orders_us.csv')

# El df visits contiene la cantidad de las visitas de cada grupo por fecha(YYYY-MM-DD).
visits = pd.read_csv('./datasets/visits_us.csv')

In [7]:
# Cambio los nombres de las columnas para que estén en minúsculas.

hypotheses.columns = hypotheses.columns.str.lower()

**Data cleaning: duplicates and missing data**

In [8]:
#Busco duplicados y valores perdidos en el df hypotheses

if hypotheses.duplicated().sum() == 0:
    print('No se encontraron filas duplicadas en este dataframe')
    
else:
    print(f'hay{hypotheses.duplicated().sum()}valores duplicados')

if hypotheses.isna().sum().sum() == 0:
    print('No se encontraron valores perdidos en este dataframe')
else:
    print(f"\nHay {hypotheses.isna().sum().sum()} valores perdidos en este dataframe")

No se encontraron filas duplicadas en este dataframe
No se encontraron valores perdidos en este dataframe


In [9]:
#Busco duplicados y valores perdidos en el df orders

if orders.duplicated().sum() == 0:
    print('No se encontraron filas duplicadas en este dataframe')
    
else:
    print(f'hay{orders.duplicated().sum()}valores duplicados')

if orders.isna().sum().sum() == 0:
    print('No se encontraron valores perdidos en este dataframe')
else:
    print(f"\nHay {orders.isna().sum().sum()} valores perdidos en este dataframe")


No se encontraron filas duplicadas en este dataframe
No se encontraron valores perdidos en este dataframe


In [10]:
#Busco duplicados y valores perdidos en el df visits

if visits.duplicated().sum() == 0:
    print('No se encontraron filas duplicadas en este dataframe')
    
else:
    print(f'hay{visits.duplicated().sum()}valores duplicados')

if visits.isna().sum().sum() == 0:
    print('No se encontraron valores perdidos en este dataframe')
else:
    print(f"\nHay {visits.isna().sum().sum()} valores perdidos en este dataframe")

No se encontraron filas duplicadas en este dataframe
No se encontraron valores perdidos en este dataframe


In [11]:
# Voy a buscar más minuciosamente, si existe algún usuario que aparezca en ambos grupos (A y B).
# Agrupo el df por id de visitante.

groups = orders.groupby('visitorId')['group'].nunique().reset_index()
duplicates =groups[groups['group'] > 1]

# Sí se encontraron varios, voy a almacenar en una serie los id de los usuarios duplicados.
duplicates_ids = duplicates['visitorId']

# Filtro el df para que sólo aparezca la información de los usuarios no duplicados.
filtered_orders = orders[~orders['visitorId'].isin(duplicates_ids)]
filtered_orders


,transactionId,visitorId,date,revenue,group
0,3667963787,3312258926,2019-08-15,30.4,B
1,2804400009,3642806036,2019-08-15,15.2,B
3,3797467345,1196621759,2019-08-15,155.1,B
4,2282983706,2322279887,2019-08-15,40.5,B
5,182168103,935554773,2019-08-15,35.0,B
...,...,...,...,...,...
1191,3592955527,608641596,2019-08-14,255.7,B
1192,2662137336,3733762160,2019-08-14,100.8,B
1193,2203539145,370388673,2019-08-14,50.1,A
1194,1807773912,573423106,2019-08-14,165.3,A


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - Tu enfoque inicial para la preparación y limpieza de datos está bien encaminado. Es positivo que estés verificando duplicados y valores perdidos, además de identificar usuarios que aparecen en ambos grupos del test A/B, ya que es un paso crítico para garantizar la validez del experimento.

Sin embargo, sería recomendable mejorar el manejo de los usuarios duplicados. En lugar de simplemente excluirlos, podrías evaluar su impacto en los datos (¿qué porcentaje representan, por ejemplo?) y decidir si es mejor excluirlos o reasignarlos de forma uniforme a un solo grupo para no reducir excesivamente el tamaño de las muestras. Esto fortalecería la confiabilidad de tus resultados.

¡Buen trabajo y sigue adelante con este enfoque minucioso!

</div>

# Parte 1 Priorizar hipótesis

Es momento de priorizar las hipótesis para averiguar cual de todas es la más viable y por lo tanto la que desarrollaremos. Voy a utilizar la metodología RICE para determinar cual es la hipótesis más viable.

In [12]:
# Creo las columnas ICE y RICE para cada una de las hipótesis.

hypotheses['ICE'] = ((hypotheses['impact'] * hypotheses['confidence']
)/hypotheses['effort']).round(2)

hypotheses['RICE'] = (
    hypotheses['reach'] * hypotheses['impact'] * hypotheses['confidence']
)/ hypotheses['effort']

hypotheses

,hypothesis,reach,impact,confidence,effort,ICE,RICE
0,Add two new channels for attracting traffic. T...,3,10,8,6,13.33,40.0
1,Launch your own delivery service. This will sh...,2,5,4,10,2.00,4.0
2,Add product recommendation blocks to the store...,8,3,7,3,7.00,56.0
3,Change the category structure. This will incre...,8,3,3,8,1.12,9.0
4,Change the background color on the main page. ...,3,1,1,1,1.00,3.0
5,Add a customer review page. This will increase...,3,2,2,3,1.33,4.0
6,Show banners with current offers and sales on ...,5,3,8,3,8.00,40.0
7,Add a subscription form to all the main pages....,10,7,8,5,11.20,112.0
8,Launch a promotion that gives users discounts ...,1,9,9,5,16.20,16.2


In [13]:
# Hipótesis con el ICE más alto

hypotheses[['hypothesis', 'reach', 'impact', 'confidence', 'effort', 'ICE']].sort_values(by='ICE', ascending=False)

,hypothesis,reach,impact,confidence,effort,ICE
8,Launch a promotion that gives users discounts ...,1,9,9,5,16.20
0,Add two new channels for attracting traffic. T...,3,10,8,6,13.33
7,Add a subscription form to all the main pages....,10,7,8,5,11.20
6,Show banners with current offers and sales on ...,5,3,8,3,8.00
2,Add product recommendation blocks to the store...,8,3,7,3,7.00
1,Launch your own delivery service. This will sh...,2,5,4,10,2.00
5,Add a customer review page. This will increase...,3,2,2,3,1.33
3,Change the category structure. This will incre...,8,3,3,8,1.12
4,Change the background color on the main page. ...,3,1,1,1,1.00


In [14]:
# Hipótesis por el valor RICE más alto.

hypotheses[['hypothesis', 'reach', 'impact', 'confidence', 'effort', 'RICE']].sort_values(by='RICE', ascending=False)

,hypothesis,reach,impact,confidence,effort,RICE
7,Add a subscription form to all the main pages....,10,7,8,5,112.0
2,Add product recommendation blocks to the store...,8,3,7,3,56.0
0,Add two new channels for attracting traffic. T...,3,10,8,6,40.0
6,Show banners with current offers and sales on ...,5,3,8,3,40.0
8,Launch a promotion that gives users discounts ...,1,9,9,5,16.2
3,Change the category structure. This will incre...,8,3,3,8,9.0
1,Launch your own delivery service. This will sh...,2,5,4,10,4.0
5,Add a customer review page. This will increase...,3,2,2,3,4.0
4,Change the background color on the main page. ...,3,1,1,1,3.0


**Conclusion de la priorización de hipótesis**

La hipótesis número 8 parecía ser la mejor con el enfoque ICE, pero el hecho que su alcance (reach) sea tan bajo (de 1) hace que se desplome y caiga muchas posiciones en el enfoque RICE, quiere decir que era una hipótesis bien elaborada, con un grado alto de confianza e impacto, pero iba a abarar a muy pocas personas.

La hipótesis número 0 es una buena prospecta, tiene valores bastante altos en ambas escalas (ICE y RICE), excelente impacto y alto grado de confianza, sin embargo su alcance también es bajo y su esfuerzo no es tan bajo, además hay una hipótesis que le supera en la escala RICE.

La hipótesis número 7, es la que tiene el valor RICE más alto por sobre todas las demás con 112 puntos, tiene el doble de puntaje que la hipótesis número 0. Y se encuentra entre las 3 hipótesis con el valor ICE más alto, es por ello que es la hipótesis que he decidido investigar.

In [15]:
print(f"La hipótesis que voy a investigar es la número 7:\n\n{hypotheses.loc[7,'hypothesis']}")

La hipótesis que voy a investigar es la número 7:

Add a subscription form to all the main pages. This will help you compile a mailing list


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - ¡Excelente trabajo! Has aplicado de manera adecuada las metodologías ICE y RICE para priorizar hipótesis. La claridad en la descripción y el análisis del impacto de los valores de alcance (reach) en las decisiones es muy destacable. Esto demuestra una comprensión sólida de cómo estas métricas pueden influir en la selección de hipótesis.

</div> 

# Parte 2. Análisis del test A/B


## Ingreso acumulado por grupo

Observemos como fueron los ingresos para cada grupo.

In [43]:
# Ingresos por grupo
revenue = orders.groupby('group')['revenue'].mean().reset_index()

fig = px.bar(
revenue, x='group', y='revenue',
width=800, height=600, color="group", title="El grupo B produjo mayores ingresos totales"
)

fig.show()

In [44]:
fig = px.histogram(
    orders, 
    x='revenue', 
    color='group',  # Diferenciar por grupo A/B
    barmode='overlay',  # Superponer las distribuciones
    marginal='box',  # Agregar un boxplot en la parte superior (opcional)
    width=800, 
    height=600, 
    title="Distribución de ingresos por grupo"
)

# Mostrar el gráfico
fig.show()

El grupo B produjo mayores ingresos, sin embargo habrá que analizar más minuciosamente, dependiendo del número de personas en cada grupo y después de eliminar valores de anómalos, podríamos tener difenrentes resultados al observar el consumo promedio de las personas de cada grupo. 

In [17]:
# Cambio las columnas date al tipo datetime 
orders['date'] = pd.to_datetime(orders['date'])
visits['date'] = pd.to_datetime(visits['date'])


In [18]:
# Métrica acumulada de transacciones, visitantes y ganancia por día para cada grupo.

datesGroups = orders[['date','group']].drop_duplicates()
ordersAggregated = datesGroups.apply(lambda x: orders[np.logical_and(orders['date'] <= x['date'], orders['group'] == x['group'])].agg({'date' : 'max', 'group' : 'max', 'transactionId' : pd.Series.nunique, 'visitorId' : pd.Series.nunique, 'revenue' : 'sum'}), axis=1).sort_values(by=['date','group'])

ordersAggregated.head()

,date,group,transactionId,visitorId,revenue
55,2019-08-01,A,24,20,2356.8
66,2019-08-01,B,21,20,1620.0
175,2019-08-02,A,44,38,3860.3
173,2019-08-02,B,45,43,4243.8
291,2019-08-03,A,68,62,5675.5


In [19]:
# Métrica acumulada de visitas por día para cada grupo.

visitorsAggregated = datesGroups.apply(lambda x: visits[np.logical_and(visits['date'] <= x['date'], visits['group'] == x['group'])].agg({'date' : 'max', 'group' : 'max', 'visits' : 'sum'}), axis=1).sort_values(by=['date','group'])
visitorsAggregated

,date,group,visits
55,2019-08-01,A,719
66,2019-08-01,B,713
175,2019-08-02,A,1338
173,2019-08-02,B,1294
291,2019-08-03,A,1845
...,...,...,...
496,2019-08-29,B,17708
730,2019-08-30,A,18037
690,2019-08-30,B,18198
958,2019-08-31,A,18736


In [20]:
#  Fusiono las dos tablas en una y coloco a sus columnas nombres descriptivos.

cumulativeData = ordersAggregated.merge(visitorsAggregated, left_on=['date', 'group'], right_on=['date', 'group'])
cumulativeData.columns = ['date', 'group', 'orders', 'buyers', 'revenue', 'visitors']

In [21]:
# Gráfica de líneas de la ganancia acumulada.

fig = px.line(
    cumulativeData, 
    x='date', 
    y='revenue', 
    color='group',  # Diferenciar los grupos con colores
    title='El grupo B tiene una mayor cantidad de ingresos (A/B Test)',
    line_shape='linear',  # Asegura líneas rectas
    markers=True  # Mostrar puntos en las líneas
)

# Personalizar la leyenda (arriba a la izquierda)
fig.update_layout(
    legend_title="Grupo",
    legend=dict(
        x=0.02,   # Posición horizontal (izquierda)
        y=0.98,   # Posición vertical (arriba)
        xanchor="left",
        yanchor="top"
    ),
    xaxis_title="Fecha", 
    yaxis_title="Ingresos Acumulados ($)",
)

fig.show()



**Ganancia acumulada de la prueba**

En el gráfico podemos observar como la ganancia acumulada es mayor en el grupo B, sin embargo valdría la pena averiguar qué pasó el 19 de agosto, parece que hubo una cantidad de compras atípica en el grupo B, si descubrimos que las compras de esa fecha pertenecen a valores anómalos, quizá la brecha entre el grupo A y grupo B podría no ser tan grande ya que los gráficos son bastante similares, parece que el día 19 de agosto hizo que subiera mucho la gráfica pero podría tratarse de un valor atípico que podríamos rechazar para tener una aproximación más precisa. 

## Tamaño de pedido promedio acumulado por grupo

In [22]:
# Creo una columna con el tamaño promedio de los pedidos para cada grupo (Revenue per Orders "rpo").

cumulativeData['rpo'] = (cumulativeData['revenue'] / cumulativeData['buyers']).round(2)

cumulativeData.head()

,date,group,orders,buyers,revenue,visitors,rpo
0,2019-08-01,A,24,20,2356.8,719,117.84
1,2019-08-01,B,21,20,1620.0,713,81.00
2,2019-08-02,A,44,38,3860.3,1338,101.59
3,2019-08-02,B,45,43,4243.8,1294,98.69
4,2019-08-03,A,68,62,5675.5,1845,91.54


In [23]:
# Gráfica de líneas para mostrar el rpo acumulado para cada grupo.

fig = px.line(
    cumulativeData, 
    x='date', 
    y='rpo', 
    color='group',  # Diferenciar los grupos por color
    labels={'date': 'Fecha', 'revenue_per_order': 'Ingresos Promedio por Pedido', 'group': 'Grupo'},
    title='Ingresos Promedio por Pedido acumulado para cada Grupo (A/B Test)',
    markers=True  # Mostrar puntos en las líneas
)

# Personalizar la leyenda (arriba a la izquierda)
fig.update_layout(
    legend_title="Grupo",
    legend=dict(
        x=0.02,   # Posición horizontal (izquierda)
        y=0.98,   # Posición vertical (arriba)
        xanchor="left",
        yanchor="top"
    ),
    xaxis_title="Fecha", 
    yaxis_title="Ingresos Promedio por Pedido ($)",
)

# Mostrar el gráfico
fig.show()

Definitivamente el 19 de agosto ocurrieron compras atípicas que están modificando nuestra percepción de los datos, tendré que hacer un filtrado de los valores atípicos antes de llegar a conclusiones.

## Diferencia relativa

In [24]:
# Separo el df por grupos 

groupA = cumulativeData[cumulativeData['group'] == 'A']
groupB = cumulativeData[cumulativeData['group'] == 'B']

# uno el df por fecha

mergedCumulativeRevenue = groupA.merge(
    groupB, 
    on='date', 
    how='left', 
    suffixes=['A', 'B']
)

# Calculo la diferencia relativa en el tamaño promedio de compra
mergedCumulativeRevenue['relative_difference'] = (
    (mergedCumulativeRevenue['revenueB'] / mergedCumulativeRevenue['ordersB']) /
    (mergedCumulativeRevenue['revenueA'] / mergedCumulativeRevenue['ordersA']) - 1
)

mergedCumulativeRevenue.head()

,date,groupA,ordersA,buyersA,revenueA,visitorsA,rpoA,groupB,ordersB,buyersB,revenueB,visitorsB,rpoB,relative_difference
0,2019-08-01,A,24,20,2356.8,719,117.84,B,21,20,1620.0,713,81.00,-0.214431
1,2019-08-02,A,44,38,3860.3,1338,101.59,B,45,43,4243.8,1294,98.69,0.074915
2,2019-08-03,A,68,62,5675.5,1845,91.54,B,61,59,6050.6,1803,102.55,0.188429
3,2019-08-04,A,84,77,6822.1,2562,88.60,B,78,75,7767.9,2573,103.57,0.226225
4,2019-08-05,A,109,100,8810.2,3318,88.10,B,101,98,9262.9,3280,94.52,0.134662


In [25]:
# Creo una gráfica para mostrarlo.

# Crear el gráfico de líneas con Plotly Express
fig = px.line(
    mergedCumulativeRevenue, 
    x='date', 
    y='relative_difference', 
    title='El grupo B tiene un mejor rendimiento en términos de ingresos promedio por pedido.',
    labels={'date': 'Fecha', 'relative_difference': 'Diferencia Relativa'},
    line_shape='linear',  # Líneas rectas
    markers=True  # Agregar puntos en la línea
)

# Agregar una línea horizontal de referencia en y=0
fig.add_hline(y=0, line_dash="dash", line_color="black", annotation_text="Referencia")

# Personalizar el diseño del gráfico
fig.update_layout(
    xaxis_title="Fecha",
    yaxis_title="Diferencia Relativa",

)

# Mostrar el gráfico
fig.show()

Este gráfico nos muestra que el grupo B tiene un mejor rendimiento, aunque me sigue causando ruido el 19 de agosto, que podría estar modificando nuestros resultados

## Relación entre pedidos y número de visitas de cada día

In [26]:
# Obtengo la tasa de conversión para ambos grupos

groupA = groupA.copy()
groupA['conversion'] = groupA['orders']/groupA['visitors']
groupB = groupB.copy()
groupB['conversion'] = groupB['orders']/groupB['visitors']


In [27]:
# Junto ambos df

groups_conversion = pd.concat([groupA, groupB,], ignore_index=True)
groups_conversion

,date,group,orders,buyers,revenue,visitors,rpo,conversion
0,2019-08-01,A,24,20,2356.8,719,117.84,0.033380
1,2019-08-02,A,44,38,3860.3,1338,101.59,0.032885
2,2019-08-03,A,68,62,5675.5,1845,91.54,0.036856
3,2019-08-04,A,84,77,6822.1,2562,88.60,0.032787
4,2019-08-05,A,109,100,8810.2,3318,88.10,0.032851
...,...,...,...,...,...,...,...,...
57,2019-08-27,B,548,504,82161.3,16523,163.02,0.033166
58,2019-08-28,B,576,526,83719.3,17177,159.16,0.033533
59,2019-08-29,B,596,544,87343.0,17708,160.56,0.033657
60,2019-08-30,B,620,567,90736.3,18198,160.03,0.034070


In [28]:
# Muestro las tasas de conversión.

fig = px.line(
    groups_conversion, x='date', y='conversion',color='group',
    title='El grupo B muestra una mayor tasa de conversión (A/B Test)',
    line_shape='linear',  # Asegura líneas rectas
    markers=True  # Mostrar puntos en las líneas
)

# Personalizar la leyenda (arriba a la izquierda)
fig.update_layout(
    legend_title="Grupo",
    legend=dict(
        x=0.02,   # Posición horizontal (izquierda)
        y=0.98,   # Posición vertical (arriba)
        xanchor="left",
        yanchor="top"
    ),
    xaxis_title="Fecha", 
    yaxis_title="Conversión",
)

fig.show()

Al comienzo los datos son inestables, pero conforme avanza el experimento la conversión en el grupo A muestra una tendencia a la baja, por el contrario el grupo B muestra tendencia al alza. Todo parece apuntar a que el grupo B es superior.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - ¡Gran trabajo en esta sección! Has presentado un análisis sólido del test A/B, incluyendo métricas acumuladas como ingresos, tamaño promedio de pedido, diferencia relativa y tasas de conversión. La inclusión de gráficos bien diseñados para visualizar tendencias y comparaciones entre grupos facilita la interpretación de los resultados. Es especialmente valioso que te hayas enfocado en identificar un posible valor atípico (el 19 de agosto), lo cual refleja un enfoque crítico en el análisis.

</div> 

## número de pedidos por usuario

Ahora estudiemos el número de pedidos que realizó cada usuario en cada grupo.

In [29]:
# Agrupo los datos por visitorId y grupo
orders_by_users = orders.groupby(['visitorId', 'group']).agg({'transactionId': 'nunique'}).reset_index()
orders_by_users.columns = ['visitorId', 'group', 'transactions']
orders_by_users

,visitorId,group,transactions
0,5114589,B,1
1,6958315,B,1
2,8300375,A,1
3,8300375,B,1
4,11685486,A,1
...,...,...,...
1084,4266935830,B,1
1085,4278982564,A,1
1086,4279090005,B,1
1087,4281247801,B,1


In [30]:
# Muestro un gráfico de dispersión con los pedidos que realizó cada grupo en cada día.

fig = px.scatter(
    orders_by_users, 
    x='visitorId',                # Eje X: Usuarios
    y='transactions',             # Eje Y: Cantidad de transacciones
    color='group',                 # Diferenciar por grupo A/B
    title='Número de transacciones por usuario',  
    labels={'visitorId': 'Usuarios', 'transactionId': 'Número de Transacciones'},
    opacity=0.7,                    # Transparencia de los puntos
    color_discrete_map={'A': '#636EFA', 'B': '#EF553B'}
)

# Mostrar el gráfico
fig.show()

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - El gráfico presentado ofrece una visualización interesante del número de transacciones por usuario. Sin embargo, usar el ID de usuario como eje X puede no ser la mejor opción para este tipo de análisis, ya que los ID son valores categóricos que no tienen un orden lógico ni un significado continuo. Esto puede dificultar la interpretación de patrones en los datos.

Una alternativa sería agrupar a los usuarios por intervalos de transacciones y crear un histograma o gráfico de barras para mostrar la distribución de las transacciones por grupo. Esto facilitaría la comparación entre los grupos A y B y destacaría diferencias clave en su comportamiento.

</div>

Parece que el número de pedidos en ambos grupos es mayormente de 1 por usuario, son pocos los usuarios que ordenan más de 2 veces.

## Definir anomalías (Pedidos por usuario al día)

Voy a estudiar los datos en los percentiles 99 y 95 para aislarlos de anomalías.

In [31]:
# Calculo los percentiles 95 y 99 de mi distribución de datos.

print("Percentiles 95 y 99:", np.percentile(orders_by_users['transactions'], [95, 99]))

Percentiles 95 y 99: [2. 4.]


Según el percentil 95 **cualquier usuario que tenga más de 2 pedidos o deberá considerarse como una anomalía**.

## Precios de los pedidos.

Observemos los precios en los pedidos de cada grupo

In [32]:
# Gráfico de dispersión para los tamaños de los pedidos (orders)

fig = px.scatter(
    orders, x='transactionId', y='revenue', color='group', title='El grupo B cuenta con 2 valores extremos'
)

fig.show()

El grupo B cuenta con 2 valores que están muy por fuera de los valores normales, son anomalías que no debemos tomar en cuenta, porque sesgan nuestros datos y pareciera que el grupo B es mejor que el A y podría no ser cierto o al menos no tan ámplia la diferencia entre ambos grupos.

## percentiles de los precios de los pedidos

In [33]:
# Calculo los percentiles 95 y 99 de mi distribución de datos (precios de los pedidos)

print("Percentiles 95 y 99:", np.percentile(orders['revenue'], [95, 99]))

Percentiles 95 y 99: [435.54  900.904]


Según el percentil 95 **todos los pedidos de más de $ 436 deben ser considerados como anomalías**, eso definitivamente cambiará nuestra percepción de los datos.

## significancia estadística de la diferencia en la conversión entre los grupos (datos en bruto)

In [34]:
# agrupo los usuarios que hicieron ordenes en A y B

ordersByUsersA = orders[orders['group']=='A'].groupby('visitorId', as_index=False).agg({'transactionId' : pd.Series.nunique})
ordersByUsersA.columns = ['userId', 'orders']

ordersByUsersB = orders[orders['group']=='B'].groupby('visitorId', as_index=False).agg({'transactionId' : pd.Series.nunique})
ordersByUsersB.columns = ['userId', 'orders']


In [35]:
# Agrego a cada grupo la cantidad de personas que entraron pero no realizaron ordenes (orders = 0)

sampleA = pd.concat([ordersByUsersA['orders'],pd.Series(0, index=np.arange(visits[visits['group']=='A']['visits'].sum() - len(ordersByUsersA['orders'])), name='orders')],axis=0)

sampleB = pd.concat([ordersByUsersB['orders'],pd.Series(0, index=np.arange(visits[visits['group']=='B']['visits'].sum() - len(ordersByUsersB['orders'])), name='orders')],axis=0)


In [36]:
# Calculo si existe diferencia estadística entre los grupos A y B.

p_value = round(stats.mannwhitneyu(sampleA, sampleB)[1], 3)

print('P-value:', p_value)

if p_value < 0.05:
    print("Existe diferencia significativa entre los grupos (P-value < 0.05)")
else:
    print("No hay diferencia significativa entre los grupos (P-value >= 0.05)")


# Calculo la diferencia relativa entre las medias de sampleB y sampleA
relative_difference = round((sampleB.mean() / sampleA.mean() - 1), 3)

print('\nComparación relativa:', relative_difference)

if relative_difference > 0:
    print(f"La media de B es mayor que A en un {relative_difference * 100}%")
else:
    print(f"La media de B es menor o igual que A en un {relative_difference * 100}%")


P-value: 0.017
Existe diferencia significativa entre los grupos (P-value < 0.05)

Comparación relativa: 0.138
La media de B es mayor que A en un 13.8%


Con los datos sin procesar tenemos una diferencia estadística que nos indica que los grupos son diferentes, y que la media del grupo B es mayor a la del A en un 13.8%

## Significancia estadística de la diferencia en el tamaño promedio de pedido entre los grupos (datos en bruto)

In [37]:
# Comparo el tamaño promedio de los pedidos para ambos grupos.

p_value = round(stats.mannwhitneyu(
    orders[orders['group'] == 'A']['revenue'], 
    orders[orders['group'] == 'B']['revenue']
)[1], 3)

# Calculo la diferencia relativa de las medias de ingresos.

relative_difference = round((orders[orders['group'] == 'B']['revenue'].mean() / \
                      orders[orders['group'] == 'A']['revenue'].mean() - 1), 3)


print(f"P-value de la prueba de Mann-Whitney U: {p_value:.3f}")

if p_value < 0.05:
    print("Existe una diferencia estadísticamente significativa entre los grupos (P-value < 0.05)")
else:
    print("No se encontró una diferencia estadísticamente significativa (P-value >= 0.05)")

print(f"\nDiferencia relativa de las medias: {relative_difference:.3f}")

if relative_difference > 0:
    print(f"El grupo B tiene un ingreso promedio {relative_difference:.1%} mayor que el grupo A.")
else:
    print(f"El grupo A tiene un ingreso promedio {-relative_difference:.1%} mayor que el grupo B.")


P-value de la prueba de Mann-Whitney U: 0.692
No se encontró una diferencia estadísticamente significativa (P-value >= 0.05)

Diferencia relativa de las medias: 0.252
El grupo B tiene un ingreso promedio 25.2% mayor que el grupo A.


Respecto a los tamaños promedios en los pedidos realizados en los grupos, la prueba (Mann-Whitney U) no encontró evidencia suficiente para decir que los grupos son estadísticamente diferentes, sin embargo la diferencia relativa de las medias indica que la media del grupo B es más del 25% mayor que la media del grupo A.

## Significancia estadística de la diferencia en la conversión entre los grupos utilizando los datos filtrados

Utilizando como base el percentil 95 vamos a quitar de los datos a los usuarios que tengan más de 2 pedidos, y a los usuarios que hayan hecho compras de más de $ 436 ya que son anomalías. 

In [38]:
# Encuentro a los usuarios con más de 2 órdenes.
usersWithManyOrders = pd.concat([ordersByUsersA[ordersByUsersA['orders'] > 2]['userId'], ordersByUsersB[ordersByUsersB['orders'] > 2]['userId']], axis = 0)

# Encuentro a los usuarios con compras de más de $436.
usersWithExpensiveOrders = orders[orders['revenue'] > 436]['visitorId']

# fusiono esas dos colecciones de ID's, y elimino duplicados.
abnormalUsers = pd.concat([usersWithManyOrders, usersWithExpensiveOrders], axis = 0).drop_duplicates().sort_values()

print(abnormalUsers.head(5))
print(abnormalUsers.shape)

1099    148427295
18      199603092
928     204675465
23      237748145
37      249864742
dtype: int64
(73,)


Tenemos 73 usuarios anómalos, vamos a eliminarlos de nuestros datos, para tener los datos límpios de esos datos extremos.

In [39]:
# Filtro los df quitando los usuarios con comportamientos anómalos.

sampleAFiltered = pd.concat([ordersByUsersA[np.logical_not(ordersByUsersA['userId'].isin(abnormalUsers))]['orders'],pd.Series(0, index=np.arange(visits[visits['group']=='A']['visits'].sum() - len(ordersByUsersA['orders'])),name='orders')],axis=0)
sampleBFiltered = pd.concat([ordersByUsersB[np.logical_not(ordersByUsersB['userId'].isin(abnormalUsers))]['orders'],pd.Series(0, index=np.arange(visits[visits['group']=='B']['visits'].sum() - len(ordersByUsersB['orders'])),name='orders')],axis=0)


In [40]:
# Calculo si existe diferencia estadística entre los grupos A y B.

p_value = round(stats.mannwhitneyu(sampleAFiltered, sampleBFiltered)[1], 3)

print('P-value:', p_value)

if p_value < 0.05:
    print("Existe diferencia significativa entre los grupos (P-value < 0.05)")
else:
    print("No hay diferencia significativa entre los grupos (P-value >= 0.05)")


# Calculo la diferencia relativa entre las medias de sampleB y sampleA
relative_difference = round((sampleBFiltered.mean() / sampleAFiltered.mean() - 1), 3)

print('\nComparación relativa:', relative_difference)

if relative_difference > 0:
    print(f"La media de B es mayor que A en un {relative_difference * 100}%")
else:
    print(f"La media de B es menor o igual que A en un {relative_difference * 100}%")

P-value: 0.014
Existe diferencia significativa entre los grupos (P-value < 0.05)

Comparación relativa: 0.17
La media de B es mayor que A en un 17.0%


Ya con los datos filtrados la diferencia estadística fue aún mayor arrojando la prueba un P-value más pequeño en esta ocación respecto del P-value que mostró con los datos sin filtrar.

Incluso ahora es mayor la diferencia entre la media de B respecto la de A.

## Significancia estadística de la diferencia en el tamaño promedio de pedido entre los grupos utilizando los datos filtrados

In [41]:
orders

,transactionId,visitorId,date,revenue,group
0,3667963787,3312258926,2019-08-15,30.4,B
1,2804400009,3642806036,2019-08-15,15.2,B
2,2961555356,4069496402,2019-08-15,10.2,A
3,3797467345,1196621759,2019-08-15,155.1,B
4,2282983706,2322279887,2019-08-15,40.5,B
...,...,...,...,...,...
1192,2662137336,3733762160,2019-08-14,100.8,B
1193,2203539145,370388673,2019-08-14,50.1,A
1194,1807773912,573423106,2019-08-14,165.3,A
1195,1947021204,1614305549,2019-08-14,5.5,A


In [42]:
# Comparo el tamaño promedio de los pedidos para ambos grupos.

p_value = round(stats.mannwhitneyu(
    orders[np.logical_and(orders['group'] == 'A', np.logical_not(orders['visitorId'].isin(abnormalUsers)))]['revenue'], 
    orders[np.logical_and(orders['group'] == 'B', np.logical_not(orders['visitorId'].isin(abnormalUsers)))]['revenue']
)[1], 3)

# Calculo la diferencia relativa de las medias de ingresos.

relative_difference = round((orders[np.logical_and(orders['group'] == 'B', np.logical_not(orders['visitorId'].isin(abnormalUsers)))]['revenue'].mean() /
                             orders[np.logical_and(orders['group'] == 'A', np.logical_not(orders['visitorId'].isin(abnormalUsers)))]['revenue'].mean() - 1), 3)


print(f"P-value de la prueba de Mann-Whitney U: {p_value}")

if p_value < 0.05:
    print("Existe una diferencia estadísticamente significativa entre los grupos (P-value < 0.05)")
else:
    print("No se encontró una diferencia estadísticamente significativa (P-value >= 0.05)")

print(f"\nDiferencia relativa de las medias: {relative_difference}")

if relative_difference > 0:
    print(f"El grupo B tiene un ingreso promedio {relative_difference:.1%} mayor que el grupo A.")
else:
    print(f"El grupo A tiene un ingreso promedio {-relative_difference:.1%} mayor que el grupo B.")

P-value de la prueba de Mann-Whitney U: 0.744
No se encontró una diferencia estadísticamente significativa (P-value >= 0.05)

Diferencia relativa de las medias: -0.028
El grupo A tiene un ingreso promedio 2.8% mayor que el grupo B.


Ya con los datos filtrados el P-value demuestra que no hay diferencia estadística como para rechazar la hipótesis nula, por lo tanto los tamaños de compra en ambos grupos son similares.

Tras esta investigación las pruebas arrojan que el los tamaños de compra son ligeramente mayores (2.8%) en el grupo A.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Éxito</b> - ¡Buen trabajo! Has realizado un análisis exhaustivo al evaluar la significancia estadística de las diferencias entre los grupos A y B tanto con datos en bruto como con datos filtrados. El uso de percentiles para identificar valores atípicos y tu enfoque metódico para filtrar usuarios anómalos muestran un enfoque crítico y cuidadoso. Además, el empleo de pruebas estadísticas como la Mann-Whitney U demuestra una comprensión sólida de cómo comparar grupos en un test A/B. También es destacable la interpretación clara y detallada de los resultados.

</div> <div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>

<b>Atención</b> ⚠️ - Una pequeña mejora que podrías implementar es incluir visualizaciones adicionales para comparar los datos antes y después de filtrar las anomalías. Por ejemplo, gráficos de caja o diagramas de dispersión podrían ayudarte a comunicar mejor cómo los valores extremos afectan los resultados. Esto hará que tu análisis sea más accesible para audiencias no técnicas y ayudará a respaldar tus conclusiones.

</div>

## Conclusiones

Después de haber estudiado los datos sin procesar y los datos filtrados, no revelaron diferencias estadísticamente significativas en la conversión entre los grupos.

Tampoco los datos sin procesar ni los filtrados revelaron diferencias estadísticamente significativas en el tamaño promedio de los pedidos entre los grupos.

Por lo tanto tomo la decisión de detener la prueba y concluyo que no existen razones para pensar que el grupo B es mejor que el grupo A a pesar de haber obtenido mayores ingresos.




# Comentario General del Revisor

<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

¡Felicidades! Tu proyecto está **aprobado**. Has demostrado un excelente manejo de los datos, un análisis estadístico robusto, y una comprensión profunda de las metodologías para tests A/B. A continuación, destaco los puntos positivos y algunas áreas interesantes para seguir investigando:

#### Puntos Positivos:
- **Limpieza de datos y manejo de anomalías:** Has identificado y filtrado valores atípicos de manera efectiva, asegurando un análisis más preciso.
- **Aplicación de pruebas estadísticas:** El uso de Mann-Whitney U y el cálculo de diferencias relativas están correctamente aplicados y bien explicados.
- **Interpretación de resultados:** Tus conclusiones están claramente articuladas y se basan en los datos, lo que refuerza la validez de tus hallazgos.
- **Visualizaciones efectivas:** Has utilizado gráficos informativos que apoyan y complementan el análisis.

#### Áreas para Seguir Investigando:
- **Optimización de visualizaciones:** Podrías explorar incluir gráficos que muestren comparaciones más claras entre datos filtrados y no filtrados, como diagramas de caja.
- **Análisis de sensibilidad:** Investigar cómo diferentes umbrales para la identificación de anomalías podrían impactar los resultados.
- **Modelos predictivos:** Considera explorar modelos avanzados para predecir tendencias de conversión o ingresos futuros basados en los datos del test A/B.
- **Segmentación adicional:** Evaluar si el comportamiento de ciertos subgrupos de usuarios (por ejemplo, por región o tipo de producto) presenta diferencias significativas.

¡Sigue así, estás haciendo un gran trabajo! 💪🚀😊

</div> 

--- 

